### PR TITLE
evaluator: Check function call min and max arguments in new evaluator.

### PR DESCRIPTION
**Pull request by yanzhe-chen**
For github issue [#758](https://github.com/pingcap/tidb/issues/758).

1. Add min and max arguments checking.
2. Fix the incorrect min/max value for `builtinDate` function call.

### CODE REVIEW

**Comment by ngaut** [comment]
LGTM @coocood PTAL

**Comment by coocood** [comment]
Thank you!
And as always, please add some test case for this change.

**Comment by coocood** [comment]
Make sure the test cases fails before this change and passes after this change.

**Comment by yanzhe-chen** [comment]
@coocood Adding the test for this is a little bit tricky since we have to fake a wrong `builtin.Func` to trigger this failure. I am trying to do it and will update this request when finished.

**Comment by yanzhe-chen** [comment]
@coocood I have added the test for this case. In order to create the fake `builtin.Func`, I have to import two extra modules (`expression/builtin` and `model`). I am not sure this is the optimal way. Please review it.

File: optimizer/evaluator/evaluator_test.go [hunk 1]
```diff
@@ -382,6 +384,30 @@ func (s *testEvaluatorSuite) TestConvert(c *C) {
...
+	}
+	ctx := mock.NewContext()
+
+	// Test case for unknown function
+	expr := &ast.FuncCallExpr{
+		FnName: model.NewCIStr("unknown"),
+		Args:   []ast.ExprNode{},
+	}
+	_, err := Eval(ctx, expr)
+	c.Assert(err, NotNil)
```
Comment (coocood) [inline optimizer/evaluator/evaluator_test.go]: You can assert equal error, like 
`c.Assert(ErrorInvalidOperation.Equals(err), IsTrue)`

**Comment by coocood** [comment]
You don't need to create a fake `builtin.Funcs`, just use the existing one will do.

File: optimizer/evaluator/evaluator_test.go (hunk 1 above)
Comment (coocood) [inline optimizer/evaluator/evaluator_test.go]: And add a valid call expression, assert the error is nil and value is expected.

**Comment by yanzhe-chen** [comment]
@coocood I have removed the fake `builtin.Funcs`. Also found a bug in handling unlimited upper bound of arguments. I fixed it and add related tests. PTAL.

**Comment by yanzhe-chen** [comment]
@coocood Opps, I forgot to test the value for valid call expressions and use `c.Assert(ErrorInvalidOperation.Equals(err), IsTrue)` to assert equal error. I will refine that.

**Comment by yanzhe-chen** [comment]
@coocood I have added the check for the return values from valid calls.

For assert equal error, I didn't found any use of `ErrorInvalidOperation.Equals` in the source code and the existing tests in `evaluator_test.go` also uses the `c.Assert(err, NotNil)` to assert errors.

So I think we'd better keep it consistent right now and replace all the old asserts to the new one in another independent commit.

Please help review it. Thanks!

**Comment by coocood** [comment]
LGTM

It's great to see that you also have comment on test cases, many of our existing test cases are lacking comments.

The old test case was written when we didn't have `terror` package, so we can't easily compare errors.
But it's OK to just test error is nil for now, we have plan to refine old tests.

Thank you again. Looking forward for more PRs from you.

**Pull request merged**